File: app/src/main/java/com/noxvision/app/hunting/calendar/HuntingSeasonData.kt
package com.noxvision.app.hunting.calendar

import java.util.Calendar

object HuntingSeasonData {
    enum class Bundesland(val displayName: String) {
        BADEN_WUERTTEMBERG("Baden-Wuerttemberg"),
        BAYERN("Bayern"),
        BERLIN("Berlin"),
        BRANDENBURG("Brandenburg"),
        BREMEN("Bremen"),
        HAMBURG("Hamburg"),
        HESSEN("Hessen"),
        MECKLENBURG_VORPOMMERN("Mecklenburg-Vorpommern"),
        NIEDERSACHSEN("Niedersachsen"),
        NORDRHEIN_WESTFALEN("Nordrhein-Westfalen"),
        RHEINLAND_PFALZ("Rheinland-Pfalz"),
        SAARLAND("Saarland"),
        SACHSEN("Sachsen"),
        SACHSEN_ANHALT("Sachsen-Anhalt"),
        SCHLESWIG_HOLSTEIN("Schleswig-Holstein"),
        THUERINGEN("Thueringen")
    }

    data class HuntingSeason(
        val wildlifeType: String,
        val gender: String?,
        val startMonth: Int,
        val startDay: Int,
        val endMonth: Int,
        val endDay: Int,
        val notes: String? = null
    ) {
        fun isInSeason(month: Int, day: Int): Boolean {
            val start = startMonth * 100 + startDay
            val end = endMonth * 100 + endDay
            val current = month * 100 + day

            return if (start <= end) {
                current in start..end
            } else {
                current >= start || current <= end
            }
        }

        fun formatSeasonPeriod(): String {
            val startMonthName = getMonthName(startMonth)
            val endMonthName = getMonthName(endMonth)
            return "$startDay. $startMonthName - $endDay. $endMonthName"
        }

        private fun getMonthName(month: Int): String {
            return when (month) {
                1 -> "Jan"
                2 -> "Feb"
                3 -> "Maerz"
                4 -> "Apr"
                5 -> "Mai"
                6 -> "Juni"
                7 -> "Juli"
                8 -> "Aug"
                9 -> "Sep"
                10 -> "Okt"
                11 -> "Nov"
                12 -> "Dez"
                else -> ""
            }
        }
    }

    // Bundesjagdzeiten (Basis fuer alle Bundeslaender)
    private val federalSeasons = listOf(
        // Rehwild
        HuntingSeason("Rehwild", "Bock", 5, 1, 10, 15),
        HuntingSeason("Rehwild", "Ricke", 9, 1, 1, 31),
        HuntingSeason("Rehwild", "Kitz", 9, 1, 2, 28),

        // Rotwild
        HuntingSeason("Rotwild", "Hirsch", 8, 1, 1, 31),
        HuntingSeason("Rotwild", "Tier", 8, 1, 1, 31),
        HuntingSeason("Rotwild", "Kalb", 8, 1, 2, 28),

        // Damwild
        HuntingSeason("Damwild", "Hirsch", 9, 1, 1, 31),
        HuntingSeason("Damwild", "Tier", 9, 1, 1, 31),
        HuntingSeason("Damwild", "Kalb", 9, 1, 2, 28),

        // Schwarzwild - ganzjaehrig
        HuntingSeason("Schwarzwild", "Keiler", 1, 1, 12, 31),
        HuntingSeason("Schwarzwild", "Bache", 1, 1, 12, 31, "Fuehrende Bachen geschont"),
        HuntingSeason("Schwarzwild", "Frischling", 1, 1, 12, 31),
        HuntingSeason("Schwarzwild", "Ueberlaeufer", 1, 1, 12, 31),

        // Muffelwild
        HuntingSeason("Muffelwild", "Widder", 8, 1, 1, 31),
        HuntingSeason("Muffelwild", "Schaf", 9, 1, 1, 31),
        HuntingSeason("Muffelwild", "Lamm", 9, 1, 1, 31),

        // Raubwild
        HuntingSeason("Raubwild", "Fuchs", 7, 16, 2, 28),
        HuntingSeason("Raubwild", "Dachs", 8, 1, 10, 31),
        HuntingSeason("Raubwild", "Waschbaer", 1, 1, 12, 31),
        HuntingSeason("Raubwild", "Marderhund", 1, 1, 12, 31),
        HuntingSeason("Raubwild", "Marder", 10, 16, 2, 28),

        // Niederwild
        HuntingSeason("Niederwild", "Hase", 10, 1, 1, 15),
        HuntingSeason("Niederwild", "Wildkaninchen", 1, 1, 12, 31),

        // Federwild
        HuntingSeason("Federwild", "Fasan", 10, 1, 1, 15),
        HuntingSeason("Federwild", "Wildente", 9, 1, 1, 15),
        HuntingSeason("Federwild", "Wildgans", 9, 1, 1, 15),
        HuntingSeason("Federwild", "Taube", 11, 1, 2, 20),
        HuntingSeason("Federwild", "Kraehe", 8, 1, 2, 20)
    )

    fun getSeasonsForBundesland(bundesland: Bundesland): List<HuntingSeason> {
        // For now, return federal seasons
        // In a complete implementation, this would have state-specific overrides
        return federalSeasons
    }

    fun getSeasonsInEffect(bundesland: Bundesland): List<HuntingSeason> {
        val calendar = Calendar.getInstance()
        val month = calendar.get(Calendar.MONTH) + 1
        val day = calendar.get(Calendar.DAY_OF_MONTH)

        return getSeasonsForBundesland(bundesland).filter { it.isInSeason(month, day) }
    }

    fun getUpcomingSeasons(bundesland: Bundesland, withinDays: Int = 30): List<Pair<HuntingSeason, Int>> {
        val calendar = Calendar.getInstance()
        val today = calendar.clone() as Calendar

        return getSeasonsForBundesland(bundesland)
            .filter { season ->
                val month = today.get(Calendar.MONTH) + 1
                val day = today.get(Calendar.DAY_OF_MONTH)
                !season.isInSeason(month, day)
            }
            .mapNotNull { season ->
                val daysUntil = calculateDaysUntilSeason(season)
                if (daysUntil in 1..withinDays) {
                    Pair(season, daysUntil)
                } else {
                    null
                }
            }
            .sortedBy { it.second }
    }

    private fun calculateDaysUntilSeason(season: HuntingSeason): Int {
        val today = Calendar.getInstance()
        val currentYear = today.get(Calendar.YEAR)

        val seasonStart = Calendar.getInstance().apply {
            set(Calendar.YEAR, currentYear)
            set(Calendar.MONTH, season.startMonth - 1)
            set(Calendar.DAY_OF_MONTH, season.startDay)
        }

        if (seasonStart.before(today)) {
            seasonStart.add(Calendar.YEAR, 1)
        }

        val diff = seasonStart.timeInMillis - today.timeInMillis
        return (diff / (1000 * 60 * 60 * 24)).toInt()
    }

    fun isWildlifeInSeason(wildlifeType: String, gender: String?, bundesland: Bundesland): Boolean {
        val calendar = Calendar.getInstance()
        val month = calendar.get(Calendar.MONTH) + 1
        val day = calendar.get(Calendar.DAY_OF_MONTH)

        return getSeasonsForBundesland(bundesland).any { season ->
            season.wildlifeType == wildlifeType &&
                    (gender == null || season.gender == null || season.gender == gender) &&
                    season.isInSeason(month, day)
        }
    }
}

File: NoxVision-iOS/NoxVision/Hunting/Calendar/HuntingSeasonData.swift
import Foundation

struct HuntingSeason: Identifiable {
    let id = UUID()
    let wildlifeType: String
    let gender: String?
    let startMonth: Int
    let startDay: Int
    let endMonth: Int
    let endDay: Int
    let bundesland: String?

    var isActive: Bool {
        let now = Date()
        let calendar = Calendar.current
        let month = calendar.component(.month, from: now)
        let day = calendar.component(.day, from: now)
        let currentDayOfYear = month * 100 + day
        let startDayOfYear = startMonth * 100 + startDay
        let endDayOfYear = endMonth * 100 + endDay

        if startDayOfYear <= endDayOfYear {
            return currentDayOfYear >= startDayOfYear && currentDayOfYear <= endDayOfYear
        } else {
            return currentDayOfYear >= startDayOfYear || currentDayOfYear <= endDayOfYear
        }
    }

    var isUpcoming: Bool {
        guard !isActive else { return false }
        if let days = daysUntilStart {
            return days > 0 && days <= 30
        }
        return false
    }

    var daysUntilStart: Int? {
        let calendar = Calendar.current
        let now = Date()
        var components = calendar.dateComponents([.year], from: now)
        components.month = startMonth
        components.day = startDay
        guard let startDate = calendar.date(from: components) else { return nil }

        var target = startDate
        if target < now {
            components.year = (components.year ?? 2026) + 1
            target = calendar.date(from: components) ?? startDate
        }

        return calendar.dateComponents([.day], from: now, to: target).day
    }

    var periodDescription: String {
        let months = ["", "Jan", "Feb", "Mär", "Apr", "Mai", "Jun", "Jul", "Aug", "Sep", "Okt", "Nov", "Dez"]
        return "\(startDay). \(months[startMonth]) – \(endDay). \(months[endMonth])"
    }
}

struct HuntingSeasonData {
    static func getSeasonsForBundesland(_ bundesland: String) -> [HuntingSeason] {
        // Federal hunting seasons (Bundesjagdzeiten) - default for all states
        var seasons: [HuntingSeason] = [
            // Rehwild
            HuntingSeason(wildlifeType: "Rehwild", gender: "Bock", startMonth: 5, startDay: 1, endMonth: 10, endDay: 15, bundesland: bundesland),
            HuntingSeason(wildlifeType: "Rehwild", gender: "Ricke", startMonth: 9, startDay: 1, endMonth: 1, endDay: 31, bundesland: bundesland),
            HuntingSeason(wildlifeType: "Rehwild", gender: "Kitz", startMonth: 9, startDay: 1, endMonth: 2, endDay: 28, bundesland: bundesland),

            // Schwarzwild
            HuntingSeason(wildlifeType: "Schwarzwild", gender: "Keiler", startMonth: 6, startDay: 16, endMonth: 1, endDay: 31, bundesland: bundesland),
            HuntingSeason(wildlifeType: "Schwarzwild", gender: "Bache", startMonth: 6, startDay: 16, endMonth: 1, endDay: 31, bundesland: bundesland),
            HuntingSeason(wildlifeType: "Schwarzwild", gender: "Frischling", startMonth: 1, startDay: 1, endMonth: 12, endDay: 31, bundesland: bundesland),

            // Rotwild
            HuntingSeason(wildlifeType: "Rotwild", gender: "Hirsch", startMonth: 8, startDay: 1, endMonth: 1, endDay: 31, bundesland: bundesland),
            HuntingSeason(wildlifeType: "Rotwild", gender: "Tier", startMonth: 8, startDay: 1, endMonth: 1, endDay: 31, bundesland: bundesland),
            HuntingSeason(wildlifeType: "Rotwild", gender: "Kalb", startMonth: 8, startDay: 1, endMonth: 2, endDay: 28, bundesland: bundesland),

            // Damwild
            HuntingSeason(wildlifeType: "Damwild", gender: "Hirsch", startMonth: 9, startDay: 1, endMonth: 1, endDay: 31, bundesland: bundesland),
            HuntingSeason(wildlifeType: "Damwild", gender: "Tier", startMonth: 9, startDay: 1, endMonth: 1, endDay: 31, bundesland: bundesland),
            HuntingSeason(wildlifeType: "Damwild", gender: "Kalb", startMonth: 9, startDay: 1, endMonth: 2, endDay: 28, bundesland: bundesland),

            // Raubwild
            HuntingSeason(wildlifeType: "Raubwild", gender: "Fuchs", startMonth: 7, startDay: 16, endMonth: 2, endDay: 28, bundesland: bundesland),
            HuntingSeason(wildlifeType: "Raubwild", gender: "Dachs", startMonth: 8, startDay: 1, endMonth: 10, endDay: 31, bundesland: bundesland),
            HuntingSeason(wildlifeType: "Raubwild", gender: "Waschbaer", startMonth: 8, startDay: 1, endMonth: 3, endDay: 15, bundesland: bundesland),
            HuntingSeason(wildlifeType: "Raubwild", gender: "Marderhund", startMonth: 9, startDay: 1, endMonth: 2, endDay: 28, bundesland: bundesland),
            HuntingSeason(wildlifeType: "Raubwild", gender: "Marder", startMonth: 10, startDay: 16, endMonth: 2, endDay: 28, bundesland: bundesland),

            // Niederwild
            HuntingSeason(wildlifeType: "Niederwild", gender: "Hase", startMonth: 10, startDay: 1, endMonth: 12, endDay: 31, bundesland: bundesland),
            HuntingSeason(wildlifeType: "Niederwild", gender: "Wildkaninchen", startMonth: 1, startDay: 1, endMonth: 12, endDay: 31, bundesland: bundesland),

            // Federwild
            HuntingSeason(wildlifeType: "Federwild", gender: "Fasan", startMonth: 10, startDay: 1, endMonth: 1, endDay: 15, bundesland: bundesland),
            HuntingSeason(wildlifeType: "Federwild", gender: "Wildente", startMonth: 9, startDay: 1, endMonth: 1, endDay: 15, bundesland: bundesland),
            HuntingSeason(wildlifeType: "Federwild", gender: "Wildgans", startMonth: 11, startDay: 1, endMonth: 1, endDay: 15, bundesland: bundesland),
            HuntingSeason(wildlifeType: "Federwild", gender: "Taube", startMonth: 11, startDay: 1, endMonth: 2, endDay: 20, bundesland: bundesland),
        ]

        return seasons
    }
}
